src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java

```diff
@@ -270,6 +337,74 @@ private static long expiresAtMillis(
         return expiresAtMillis;
     }
 
+    private final class OAuthAccessGrant
+    {
+        private String subject;
+        private long authorization;
+        private long expiresAt;
+        private int referenceCount;
+        private Consumer<String> cleaner;
+
+        private OAuthAccessGrant(
+            Consumer<String> cleaner)
+        {
+            this.cleaner = cleaner;
+        }
+
+        private boolean reauthorize(
+            String subject,
+            long connectAuthorization,
+            long expiresAtMillis)
+        {
+            final boolean reauthorized;
+            if (referenceCount > 0)
+            {
+                final long grantAuthorization = authorization;
+                reauthorized = (grantAuthorization & connectAuthorization) == grantAuthorization && expiresAtMillis > expiresAt;
+
+                if (reauthorized)
+                {
+                    expiresAt = expiresAtMillis;
+                }
+            }
+            else
+            {
+                this.subject = subject != null ? subject.intern() : null;
+                this.authorization = connectAuthorization;
+                this.expiresAt = expiresAtMillis;
+                reauthorized = false;
+            }
+            return reauthorized;
+        }
+
+        private void acquire()
+        {
+            assert (cleaner != null);
+            referenceCount++;
+        }
+
+        private void release()
+        {
+            assert (cleaner != null && referenceCount > 0);
+            referenceCount--;
+            if (referenceCount == 0)
+            {
+                if (subject != null)
+                {
+                    cleaner.accept(subject.intern());
```

Comment:
No need to re-intern `subject` since it was already interned before assignment, right?